IBX-4625: TaxonomyEntry/Assignment fields description added in Developer-Documentation (#1841)
ezsystems/developer-documentation: docs/content_management/field_types/field_type_reference/taxonomyentryassignmentfield.md

# TaxonomyEntryAssignment Field Type

`TaxonomyEntryAssignment` Field is used to integrate content with Taxonomy module. It allows you to select tags or categories and assign them to the content. This Field Type actually assigns tags to the content in the data action, so then you can use TaxonomyService on this content. To be able to assign tags to the content, first, you need to add a `TaxonomyEntryAssignment` Field to the Content Type definition.

| Name                     | Internal name                    | Expected input |
|--------------------------|----------------------------------|----------------|
| `TaxonomyEntryAssignment`| `ibexa_taxonomy_entry_assignment`| array with `taxonomyEntries` and `taxonomy` keys|

## PHP API Field Type 

### Input expectations

> | Type | Description |
> |------|-------------|
> |`taxonomyEntries`| `\Ibexa\Contracts\Taxonomy\Value\TaxonomyEntry` objects|
> |`taxonomy`|`string`|

Reviewer: ````suggestion
| Type     | Description | Example         |
|--------|-----------------|-----------------|
| `array` | array with `Ibexa\Contracts\Taxonomy\Value\TaxonomyEntry` objects under `taxonomy_entries` key and Taxonomy identifier under `taxonomy` key | see below. |

Example using `Ibexa\Taxonomy\FieldType\TaxonomyEntryAssignment\Value` object:

``` php
$taxonomyEntry1 = $this->taxonomyService->loadEntryByIdentifier('example_entry', 'tags');
$taxonomyEntry2 = $this->taxonomyService->loadEntryByIdentifier('example_entry_2', 'tags');

new \Ibexa\Taxonomy\FieldType\TaxonomyEntryAssignment\Value(
    [
        $taxonomyEntry1,
        $taxonomyEntry2,
        // ...
    ],
    'tags', 
);
```

Example using array:

``` php
[
    'taxonomy_entries' => [$taxonomyEntry, $taxonomyEntry2], // load entries using TaxonomyService
    'taxonomy' => 'tags',
]
```
````

# TaxonomyEntryAssignment Field Type

> `TaxonomyEntryAssignment` Field is used to integrate content with Taxonomy module. It allows you to select tags or categories and assign them to the content. This Field Type actually assigns tags to the content in the data action, so then you can use TaxonomyService on this content. To be able to assign tags to the content, first, you need to add a `TaxonomyEntryAssignment` Field to the Content Type definition.

Reviewer 2: ```suggestion
`TaxonomyEntryAssignment` Field is used to integrate content with the Taxonomy module.
It allows you to select tags or categories and assign them to content.
This Field Type assigns tags to the content in the data action, so then you can use `TaxonomyService` on this Content item.

To be able to assign tags to the content, first, you need to add a `TaxonomyEntryAssignment` Field to the Content Type definition.
```

| Name                     | Internal name                    | Expected input |
|--------------------------|----------------------------------|----------------|
| `TaxonomyEntryAssignment`| `ibexa_taxonomy_entry_assignment`| array with `taxonomyEntries` and `taxonomy` keys|

## PHP API Field Type 

### Input expectations

| Type     | Description | Example         |
|--------|-----------------|-----------------|
| `array` | array with `Ibexa\Contracts\Taxonomy\Value\TaxonomyEntry` objects under `taxonomy_entries` key and Taxonomy identifier under `taxonomy` key | see below |

> Example using `Ibexa\Taxonomy\FieldType\TaxonomyEntryAssignment\Value` object:

Reviewer 2: ```suggestion
Example using a `Ibexa\Taxonomy\FieldType\TaxonomyEntryAssignment\Value` object:

```

``` php
$taxonomyEntry1 = $this->taxonomyService->loadEntryByIdentifier('example_entry', 'tags');
$taxonomyEntry2 = $this->taxonomyService->loadEntryByIdentifier('example_entry_2', 'tags');
new \Ibexa\Taxonomy\FieldType\TaxonomyEntryAssignment\Value(
    [
        $taxonomyEntry1,
        $taxonomyEntry2,
        // ...
    ],
    'tags', 
);
```

Example using array:
``` php
[
    'taxonomy_entries' => [$taxonomyEntry, $taxonomyEntry2], // load entries using TaxonomyService
    'taxonomy' => 'tags',
]
```

### Value object

#### Properties

|Property|Type|Description|
|--------|----|-----------|

> |`$taxonomyEntry`|array of `Ibexa\Contracts\Taxonomy\Value\TaxonomyEntry`|Stores selected taxonomy entry.|

Reviewer 2: ```suggestion
|`taxonomyEntry`|array of `Ibexa\Contracts\Taxonomy\Value\TaxonomyEntry`|Stores selected taxonomy entry.|
```
By convention we omit the `$` when mentioning properties in regular text, see for example https://doc.ibexa.co/en/master/content_management/field_types/field_type_reference/imagefield/#properties

> |`$taxonomy`|`string`|Stores taxonomy identifier, all `&taxonomyEntries` have to be assigned to this taxonomy and the identifier has to match to the settings of the Field Type in Content Type configuration.|

Reviewer 2: ```suggestion
|`taxonomy`|`string`|Stores the taxonomy identifier, all `taxonomyEntries` have to be assigned to this taxonomy and the identifier has to match the settings of the Field Type in Content Type configuration.|
```

#### Constructor

> Constructor accepts `$taxonomyEntries` and `$taxonomy` as described above.

Reviewer 2: ```suggestion
The constructor accepts `taxonomyEntries` and `taxonomy` as described above.
```

#### String representation

> If has no entries: empty string

Reviewer 2: ```suggestion
If the Field has no entries - empty string.

```

> With entries: "Cars and 5 more" - a string displaying first taxonomy entry and number of rest of the entries

Reviewer 2: ```suggestion
If the Field has entries (for example: "Cars and 5 more") - a string displaying the first taxonomy entry and the number of the rest of the entries.
```
Not sure if the example is needed here at all.

#### Hash format

An array of:

- `taxonomy_entries` with numerical IDs of entries.
- `taxonomy` string identifier of a taxonomy.

#### Validation

> FieldType validates if all Taxonomy Entries from the Value are assigned to configured taxonomy.

Reviewer 2: ```suggestion
The Field Type validates if all Taxonomy Entries from the value are assigned to the configured taxonomy.
```

|Property|Type|Description|
|--------|----|-----------|
|`$taxonomyEntry`|array of `Ibexa\Contracts\Taxonomy\Value\TaxonomyEntry`|Stores selected taxonomy entry.|

> |`$taxonomy`|`string`|Stores taxonomy identifier, all `&taxonomyEntries` have to be assigned to this taxonomy and the identifier has to match to the settings of the Field Type in Content Type configuration.|

Reviewer: `&taxonomyEntries` => `$taxonomyEntries`
